File: example/javalib/config/7-web-server/test/src/WebServerTests.java
package example;
import org.junit.jupiter.api.*;
import okhttp3.*;
import org.springframework.boot.SpringApplication;
import org.springframework.context.ConfigurableApplicationContext;

import static org.junit.jupiter.api.Assertions.*;

public class WebServerTests {
  private static ConfigurableApplicationContext server;
  private static final OkHttpClient client = new OkHttpClient();

  @BeforeAll
  public static void startServer() {
    server = SpringApplication.run(WebServer.class, new String[]{});
  }

  @AfterAll
  public static void stopServer() {
    if (server != null) {
      server.close();
    }
  }

  @Test
  public void testReverseString() throws Exception {
    RequestBody body = RequestBody.create(
      "hello world",
      MediaType.parse("text/plain")
    );

    Request request = new Request.Builder()
      .url("http://localhost:8080/reverse-string")
      .post(body)
      .build();

    try (Response response = client.newCall(request).execute()) {
      assertTrue(response.isSuccessful());
      assertEquals("dlrow olleh", response.body().string());
    }
  }
}

File: example/scalalib/config/7-web-example/test/src/WebServerTests.scala
package example
import utest.*

object WebServerTests extends TestSuite {
  def withServer[T](example: cask.main.Main)(block: => T): T = {
    val server = io.undertow.Undertow.builder
      .addHttpListener(8080, "localhost")
      .setHandler(example.defaultHandler)
      .build
    server.start()

    try block
    finally server.stop()
  }

  def tests = Tests {
    test("reverseString") - withServer(WebServer) {
      val response = requests.post(
        "http://localhost:8080/reverse-string",
        data = "helloworld"
      )
      assert(response.text() == "dlrowolleh")
    }
  }
}
